core/src/main/java/aima/core/search/basic/support/BasicBidirectionalActions.java

```diff
@@ -0,0 +1,50 @@
+package aima.core.search.basic.support;
+
+import aima.core.search.api.BidirectionalActions;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author manthan.
+ */
+public class BasicBidirectionalActions<A, S> implements BidirectionalActions<A> {
+    private List<A> fromInitialStatePartList = new ArrayList<>();
+    private List<A> fromGoalStatePartList = new ArrayList<>();
+    private List<A> fromInitialStateToGoalStateList = new ArrayList<>();
+    private List<A> fromGoalStateToInitialStateList = new ArrayList<>();
+
+    public BasicBidirectionalActions(List<A> fromInitialStatePartList, List<A> fromGoalStatePartList) {
+        this.fromInitialStatePartList = fromInitialStatePartList;
+        this.fromGoalStatePartList = fromGoalStatePartList;
+    }
+
+    @Override
+    public List<A> fromInitialStatePart() {
+        return fromInitialStatePartList;
+    }
+
+    @Override
+    public List<A> fromGoalStatePart() {
+        return fromGoalStatePartList;
+    }
+
+    @Override
+    public List<A> fromInitialStateToGoalState() {
+        List<A> sol = fromGoalStatePart();
+        Collections.reverse(sol);
+        fromInitialStateToGoalStateList.addAll(fromInitialStatePart());
+        fromInitialStateToGoalStateList.addAll(sol);
+        return fromInitialStateToGoalStateList;
+    }
```

Comment:
What would be returned if this method was called twice in a row. I think it would double the size of the list of actions returned. Setting up fromInitialStateToGoalStateList and fromGoalStateToInitialStateList should only need to happen once in the constructor or a special/private init method.